tensorflow/core/kernels/training_ops.cc

```diff
@@ -310,21 +311,62 @@ struct ApplyAdamNonCuda {
                   typename TTypes<T>::ConstScalar beta2,
                   typename TTypes<T>::ConstScalar epsilon,
                   typename TTypes<T>::ConstFlat grad, bool use_nesterov) {
+    // Get params length and check if they can be vectorized by packet size.
+    Index length = var.size();
+    Index size = Eigen::internal::packet_traits<T>::size;
+    if (length % size == 0) {
+      length = length / size;
+    } else {
+      size = 1;
+    }
```

Comment:
There is no need to divide the input size by the packet size, and do "manual vectorization". If it's desirable to have shard size `(end-begin)` to be a multiple of a packet size, you can pass `block_align` to `parallelFor` (see https://bitbucket.org/eigen/eigen/src/4b28c8008901c6d760f48f26ee2e3423fd8a2b40/unsupported/Eigen/CXX11/src/Tensor/TensorDeviceThreadPool.h#lines-185). \

I think this should work:
```
[](Index index) -> Index { return Eigen::divup(index, packet_size); }
```

Reply:
I got some question when try to use this function, please see my comment below.